### PR TITLE
Make ServiceBusScaleMonitor Public for Use in Scale Controller

### PR DESCRIPTION
**Pull request by chiangvincent**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

# Overview
Currently, code in the Scale Controller and Extensions is duplicated (code for `GetMetricsAsync` and `GetScaleStatus` exist in the ActivationTask and ScalerBase classes in Scale Controller, respectively). We are currently redesigning the scale controller, and the first phase is to re-platform the scale controller to load the ServiceBusSDK. This allows us to directly instantiate + call into `ServiceBusScaleMonitor` methods. This also unifies any logical implementations that have diverged over time between runtime driven scaling (handled by extensions), and regular scaling (handled by scale controller). For example, a we had a recent CRI where `ServiceBusScaleMonitor` did not support burst scaling, but scale controller did. See [open issue.](https://github.com/Azure/azure-sdk-for-net/issues/29536)

# Testing
After making the methods public, we generated local nuget packages which we referenced in our redesign (see redesign repo [here](https://github.com/chiangvincent/ScaleController/tree/main/ScaleController/nuget))

### CODE REVIEW

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs [hunk 1]
```diff
@@ -35,7 +35,7 @@ internal class ServiceBusClientFactory
             logForwarder.Start();
         }
 
-        internal ServiceBusClient CreateClientFromSetting(string connection)
+        public ServiceBusClient CreateClientFromSetting(string connection)
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs:38]: Are you going to be calling this from ScaleController? Because it seems that ServiceBusScaleMonitor only calls CreateAdministrativeClient internally - I would have thought that ScaleController would only need the APIs provided by the monitor.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@
 
 namespace Microsoft.Azure.WebJobs.Extensions.ServiceBus.Config
 {
-    internal class ServiceBusClientFactory
+    public class ServiceBusClientFactory
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs:17]: For any new public types/methods, we need to have good xml doc.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs:38]: +1 on wondering why this needs to be public.

Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs:38]: This was needed because the instantiation of a ServiceBusScaleMonitor requires a ServiceBusReceiver, which we created through a ServiceBusClient. See code pointer [here](https://github.com/chiangvincent/ScaleController/blob/user/vchiang/msiIntegration/ScaleController/Triggers/ServiceBusTrigger.cs#:~:text=public%20override%20IScaleMonitor%20GetScaleMonitor()). Not sure if this is the best way to do this -- open to suggestions here

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs:38]: I'd rather reduce the number of dependencies that the ServiceBusScaleMonitor requires. It looks like it basically needs the IConfiguration instance and then it could handle everything else internally, i.e. creating the ClientFactory, Receiver, AdminClient. We may still need to support the old flow where the ClientFactory is injected in the extensions, but that could be left as an internal constructor. For the public constructor, we should try to reduce dependencies.

**Comment by JoshLove-msft** [comment]
We should make sure to address https://github.com/Azure/azure-sdk-for-net/issues/29536 before moving ahead with the switch in the scale controller.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs (hunk 1 above)
Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs]: Created new public constructor for ServiceBusScaleMonitor, and removed changes for ServiceBusClientFactory (based on offline conversation). Let me know if this looks good!

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs (hunk 2 above)
Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs]: Keeping the ServiceBusClientFactory internal, as discussed with @JoshLove-msft offline

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 namespace Microsoft.Azure.WebJobs.ServiceBus.Listeners
 {
-    internal class ServiceBusTriggerMetrics : ScaleMetrics
+    public class ServiceBusTriggerMetrics : ScaleMetrics
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs:11]: Need class level docs.

Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs:14]: Done

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 4]
```diff
@@ -42,6 +46,21 @@ public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBus
             _nextWarningTime = DateTime.UtcNow;
         }
 
+        public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBusEntityType, string entityPath, string connection, ILoggerFactory loggerFactory, IConfiguration configuration, AzureComponentFactory componentFactory, MessagingProvider messagingProvider, AzureEventSourceLogForwarder logForwarder, IOptions<ServiceBusOptions> options)
```
[the remark this answers is not in the document]
Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:65]: Done

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 5]
```diff
@@ -14,7 +17,7 @@
 
 namespace Microsoft.Azure.WebJobs.ServiceBus.Listeners
 {
-    internal class ServiceBusScaleMonitor : IScaleMonitor<ServiceBusTriggerMetrics>
+    public class ServiceBusScaleMonitor : IScaleMonitor<ServiceBusTriggerMetrics>
```
[the remark this answers is not in the document]
Comment (chiangvincent) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:23]: Done

**Comment by JoshLove-msft** [comment]
You'll need to regenerate and commit the API surface area. See https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#public-api-additions.

**Review by JoshLove-msft — APPROVED** [review]

**Comment by chiangvincent** [comment]
Waiting to merge in Alexey's PR at https://github.com/Azure/azure-sdk-for-net/pull/30822/files#diff-e2d09bf3bc95b6b03e0ddc0f67ea411f770344d9f1f598cd2dd9fad0bd0aa847 before we merge this change

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 6]
```diff
@@ -42,20 +49,52 @@ public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBus
...
+            _serviceBusEntityType = serviceBusEntityType;
+            _entityPath = entityPath;
+            _scaleMonitorDescriptor = new ScaleMonitorDescriptor($"{_functionId}-ServiceBusTrigger-{_entityPath}".ToLower(CultureInfo.InvariantCulture));
+            _isListeningOnDeadLetterQueue = entityPath.EndsWith(DeadLetterQueuePath, StringComparison.OrdinalIgnoreCase);
+            _logger = loggerFactory.CreateLogger<ServiceBusScaleMonitor>();
+            _nextWarningTime = DateTime.UtcNow;
+            _clientFactory = new Lazy<ServiceBusClientFactory>(() => new ServiceBusClientFactory(configuration, componentFactory, messagingProvider, logForwarder, options));
+            _administrationClient = new Lazy<ServiceBusAdministrationClient>(() => _clientFactory.Value.CreateAdministrationClient(connection));
+            _client = new Lazy<ServiceBusClient>(() => _clientFactory.Value.CreateClientFromSetting(connection));
+            _receiver = new Lazy<ServiceBusReceiver>(() => _client.Value.CreateReceiver(entityPath));
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:77]: Looks like this receiver is being created differently than the previous code [here](https://github.com/Azure/azure-sdk-for-net/blob/8947cc941137289363569dbeef191cb0e7b911e4/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs#L78) which was going through the MessagingProvider to create a batch receiver, and was also ensuring ServiceBusOptions are applied. Why the difference?

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:77]: Good catch, we should make sure the receivers are created in the same way in both places. The point of this second constructor is that we can avoid making ServiceBusClientFactory public.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs [hunk 7]
```diff
@@ -8,22 +8,25 @@
...
+    /// <summary>
+    /// Trigger metrics class for Service Bus. Used to make scaling decisions in <see cref="ServiceBusScaleMonitor"/>.
+    /// </summary>
+    public class ServiceBusTriggerMetrics : ScaleMetrics
     {
         /// <summary>
         /// The number of messages currently in the queue/topic.
         /// </summary>
-        public long MessageCount { get; set; }
+        public long MessageCount { get; internal set; }
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs:19]: I think the setters should be public. Kind of strange (and not testable) to have a public data type like this that you can instantiate but not provide values for. Also note that the base class property has a public setter.

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs:19]: Normally what we do for output types like this is make the constructor internal and provide a model factory for testing. I missed that the base class already has a setter, so I would agree that we can just revert this change.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 8]
```diff
@@ -42,20 +49,52 @@ public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBus
...
+        }
+
+        ScaleMonitorDescriptor IScaleMonitor.Descriptor
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:80]: Why the change here from public to explicit interface implementation? Can't it remain public?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 9]
```diff
@@ -42,20 +49,52 @@ public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBus
...
             }
         }
-
         async Task<ScaleMetrics> IScaleMonitor.GetMetricsAsync()
         {
             return await GetMetricsAsync().ConfigureAwait(false);
         }
 
-        public async Task<ServiceBusTriggerMetrics> GetMetricsAsync()
+        async Task<ServiceBusTriggerMetrics> IScaleMonitor<ServiceBusTriggerMetrics>.GetMetricsAsync()
```
Comment (mathewc) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:92]: Same question here on explicit interface impl. Can't it stay public?

**Comment by JoshLove-msft** [comment]
> Waiting to merge in Alexey's PR at https://github.com/Azure/azure-sdk-for-net/pull/30822/files#diff-e2d09bf3bc95b6b03e0ddc0f67ea411f770344d9f1f598cd2dd9fad0bd0aa847 before we merge this change

How does Alexey's PR fit with these changes? It looks like we are changing the interface that we implement there. Are we going to reflect that in this PR?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs [hunk 10]
```diff
@@ -111,6 +111,25 @@ public partial class SessionMessageProcessor
...
+        Microsoft.Azure.WebJobs.Host.Scale.ScaleStatus Microsoft.Azure.WebJobs.Host.Scale.IScaleMonitor.GetScaleStatus(Microsoft.Azure.WebJobs.Host.Scale.ScaleStatusContext context) { throw null; }
+        System.Threading.Tasks.Task<Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusTriggerMetrics> Microsoft.Azure.WebJobs.Host.Scale.IScaleMonitor<Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusTriggerMetrics>.GetMetricsAsync() { throw null; }
+        Microsoft.Azure.WebJobs.Host.Scale.ScaleStatus Microsoft.Azure.WebJobs.Host.Scale.IScaleMonitor<Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusTriggerMetrics>.GetScaleStatus(Microsoft.Azure.WebJobs.Host.Scale.ScaleStatusContext<Microsoft.Azure.WebJobs.ServiceBus.Listeners.ServiceBusTriggerMetrics> context) { throw null; }
+    }
+    public partial class ServiceBusTriggerMetrics : Microsoft.Azure.WebJobs.Host.Scale.ScaleMetrics
+    {
+        public ServiceBusTriggerMetrics() { }
+        public long MessageCount { get { throw null; } }
+        public int PartitionCount { get { throw null; } }
+        public System.TimeSpan QueueTime { get { throw null; } }
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs:130]: We may want to think about a better name for this property. What do you think of TimeOnQueue?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs [hunk 11]
```diff
@@ -42,20 +49,52 @@ public ServiceBusScaleMonitor(string functionId, ServiceBusEntityType serviceBus
             _nextWarningTime = DateTime.UtcNow;
         }
 
-        public ScaleMonitorDescriptor Descriptor
+        /// <summary>
+        /// Public constructor used to retrieve metrics and make scale decisions for Service Bus entities.
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs:53]:
```suggestion
        /// Initializes an instance of ServiceBusScaleMonitor, which is used to retrieve metrics and make scale decisions for Service Bus entities.
```

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs [hunk 12]
```diff
@@ -8,22 +8,25 @@
...
 
         /// <summary>
         /// The number of partitions.
         /// </summary>
-        public int PartitionCount { get; set; }
+        public int PartitionCount { get; internal set; }
 
         /// <summary>
         /// The length of time the next message has been
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTriggerMetrics.cs:27]:
```suggestion
        /// The length of time the top message on the queue has been enqueued.
```

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs (hunk 10 above)
Comment (mathewc) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs:130]: There would be some complexity in changing these names at this point, since this is a serialization contract. These values are persisted and serialized/deserialized by hosts (e.g. Functions Host) and passed back into the monitor via GetScaleStatus.

Comment (JoshLove-msft) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/api/Microsoft.Azure.WebJobs.Extensions.ServiceBus.netstandard2.0.cs:130]: Ok, let's see if we can get this approved with the current name.
